app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -1,32 +1,97 @@
+import com.lightricks.feedexercise.R
 import com.lightricks.feedexercise.data.FeedItem
+import com.lightricks.feedexercise.data.FeedRepositoryImpl
+import com.lightricks.feedexercise.data.FeedRepository
+import com.lightricks.feedexercise.network.FeedApiService
 import com.lightricks.feedexercise.util.Event
-import java.lang.IllegalArgumentException
+import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
+import io.reactivex.rxjava3.disposables.Disposable
 
 /**
  * This view model manages the data for [FeedFragment].
  */
-open class FeedViewModel : ViewModel() {
-    private val isLoading = MutableLiveData<Boolean>()
-    private val isEmpty = MutableLiveData<Boolean>()
-    private val feedItems = MediatorLiveData<List<FeedItem>>()
+open class FeedViewModel(private val feedRepository: FeedRepository) : ViewModel() {
+
+    private val stateLiveData = MutableLiveData<State>()
     private val networkErrorEvent = MutableLiveData<Event<String>>()
+    private var prevTask: Disposable? = null
+    private lateinit var getFeedFromRepoStream: Disposable
+
+    fun getIsLoading(): LiveData<Boolean> = Transformations.map(stateLiveData) { it.isLoading }
+    fun getIsEmpty(): LiveData<Boolean> = Transformations.map(stateLiveData) { it.isFeedEmpty() }
+    fun getFeedItems(): LiveData<List<FeedItem>> =
+        Transformations.map(stateLiveData) { it.feedItems }
 
-    fun getIsLoading(): LiveData<Boolean> = isLoading
-    fun getIsEmpty(): LiveData<Boolean> = isEmpty
-    fun getFeedItems(): LiveData<List<FeedItem>> = feedItems
     fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
 
     init {
+        observeRepository()
         refresh()
     }
 
     fun refresh() {
-        //todo: fix the implementation
-        isLoading.value = false
-        isEmpty.value = true
+        setStartLoading()
+        disposePreviousTask()
+        prevTask = feedRepository.fetchFeed()
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe({ setStoppedLoading() },
+                { error ->
+                    handleNetworkError(error)
+                    setStoppedLoading()
+                })
+    }
+
+    private fun disposePreviousTask() {
+        prevTask?.dispose()
+    }
+
+    private fun handleNetworkError(error: Throwable?) {
+        networkErrorEvent.value = Event(R.string.repository_network_error.toString())
+        Log.e(TAG, "Failed to fetch feed from network", error)
+    }
+
+    private fun observeRepository() {
+        getFeedFromRepoStream = feedRepository.getFeedItems().subscribe({ items ->
+            stateLiveData.value = when (stateLiveData.value) {
+                null -> State(feedItems = items)
+                else -> stateLiveData.value!!.copy(feedItems = items)
+            }
+        }, { error -> Log.e(TAG, "Failed to get feed from repository", error) })
+    }
+
+    private fun setStartLoading() {
+        stateLiveData.value?.isLoading = true
```

Comment:
That is not the way we should update the value. Like now the observers are not updated with the new value since the update is not made via the live data. The second thing is that in general we want to work we immutable objects so we don't have to synchronise it with the others thread. So what I can suggest you is to replace/make this field immutable. And there, publishing a new model with the new loading status.

```kotlin
stateLiveData.value = stateLiveData.value!!.copy(isLoading = true)
```